### DISCUSSION TITLE
Can't wrap my head around groups and playing only on selected devices

### DISCUSSION THREAD

**Discussion by hillbicks** [snapcast/snapcast#828] [Q&A]
Hey guys,

hopefully someone can help me out here. I started my journey with snapcast with a really minimalistic setup, so I could get the hang of how snapcast works and how it will fit my requirements. So here is what I did and where I have trouble understanding some functionality. 

### snapcast server
Set up a dedicated lxc container with a snapcast server. The only change I made in the config file, is giving the only source stream a different name: 

`source = pipe:///tmp/snapfifo?name=mopidy`

### snapcast clients
In order to keep it really simple, I just used an android tablet and my android phone with the snapdroid clients. 

### Starting the server and both clients
The clients are connected to the server and are showing up in the app. I can do  cat /dev/urandom > /tmp/snapfifo and the noise is played on both devices. Simple enough and it works, but it also raises some questions.

### My understanding of how snapcast works: 
Groups are created by or for the individual streams defined in the server, in my case the source is tmp/snapfifo with the name mopidy. I can then pipe whatever audio output I have into this pipe, snapserver distributes that to connected clients. 

### Problem description
Now, doing so results in two groups within the snapdroid app with each client under one group. Both groups are named mopidy and both clients are already playing the noise from /dev/urandom.  I can open the settings for the first group and add the second client to this group as well, resulting in only one group. But so far, the only difference that seems to make is that I can change the volume for the group and therefore, for both clients. But is there any other advantage? Is there not a way to stop playback for one client? Or can I only mute the client in order to "stop" playback? Or stop the snapcast client where playback should be stopped?

I saw in the documentation for the API of snapcast, that I can join or unjoin clients for a group. But like I mentioned earlier, removing one of the clients from the group just results in two group entries with one client in it. They both are playing the same output from /dev/urandom. 

### What I want to achieve as a starting point:
 Pipe /dev/urandom into /tmp/snapfifo and just play it on one device. I can achieve that by stopping the snapclient but that doesn't seem very practical.

So, how do I do achieve: Start playback on the server, have two clients running, but only play back on one device. Or to put it differently: Sometimes I want to playback music in the kitchen, sometimes in the living room, sometimes on both at the same time. 

To make matters worse, there is a blog entry from home assistant (granted, from 2016) that has [each room listed with both a snapcast client AND a mopidy instance](https://www.home-assistant.io/images/blog/2016-02-snapcast/diagram.png) which confuses me even more. 

EDIT: [Link to article](https://www.home-assistant.io/blog/2016/02/18/multi-room-audio-with-snapcast/)

I'm pretty sure I'm being really stupid here and not seeing the obvious, so I would appreciate a hint in the right direction, but I couldn't really find an answer to my question, or maybe I didn't look in the right places.

**Comment by kingosticks** [comment]
> To make matters worse, there is a blog entry from home assistant (granted, from 2016) that has each room listed with both a snapcast client AND a mopidy instance which confuses me even more.

I'm not familiar with this article but I would guess those extra Mopidy instances mean that you can have each room choose to either use its Snapclient output (i.e. music coming from the shared "server" Mopidy instance), or its Mopidy output (i.e. music coming from the room's local Mopidy instance). This would allow you to listen to different songs in each room at the same time, you cannot do that with just a single shared Mopidy instance on the "server".

**Reply by hillbicks** [reply]
Thanks for your answer first of all. 

I have to think about that a bit more. For the use case of playing different streams in different rooms, I think that would work. Have a pi in each room and every pi is running it's own mopidy instance, which I then can control either through that mopidy instance, or through home assistant by setting up each mopidy instance. 

That still doesn't solve my multiroom audio problem though. What it boils down to: snapcast streams one stream (spotify, mopidy, etc) to ALL snapcast clients connected to the server. The only way around that, is to either mute the clients where I don't want the stream, or use smart electrical outlets to turn on/off the speakers, depending on the use case. 

Would leave me with the disadvantage of not being able to play radio in the kitchen and spotify in the rest of the house, although I'm not even sure that would be a problem. Currently it is either play radio in the kitchen, downstairs or the whole house, but not a combination. 

What I really don't understand: I really can't be the only one that wants to have more granularity with the multiroom setup of snapcast than: Play on all devices or none. I'm not bashing @badaix in any way, this is a hobby project and a damn good one at that. I just want to understand how other people have solved this issue, so I don't have to reinvent the wheel here.

**Reply by kingosticks** [reply]
To be clear, I was only trying to answer why someone might have both a Snapcast client *and* a Mopidy instance running on each "room" device. A similar thing could be achieved with multiple Mopidy instances running on the "server", each providing a stream to the single Snapserver.

I'm still a bit unsure what your actual issue is. 

> Pipe /dev/urandom into /tmp/snapfifo and just play it on one device. I can achieve that by stopping the snapclient but that doesn't seem very practical.

Or using snapdroid (or whatever controller method) to mute the device you do not want to hear? What is wrong with doing that?

> What it boils down to: snapcast streams one stream (spotify, mopidy, etc) to ALL snapcast clients connected to the server. 

No. Snapcast will provide as many streams as you configure either in the snapserver config or using JSON-RPC. When using Mopidy, each Mopidy instance provides one stream. Each of which is configured as a separate snapserver source. So if you want want snapcast to provide two streams,  you need two instances of Mopidy running somewhere, each of which is configured as a separate source for your snapserver instance. You can have a Mopidy-A playing some online radio and Mopidy-B playing some Spotify playlist and then use a controller (e.g. snapdroid) to select which snapclient is playing which of the two streams, or have them both play the same, or mute some/all of them. 

What granularity is this missing? Am I am misunderstanding the question?

**Reply by kingosticks** [reply]
> No. Snapcast will provide as many streams as you configure either in the snapserver config or using JSON-RPC. When using Mopidy, each Mopidy instance provides one stream. Each of which is configured as a separate snapserver source. So if you want want snapcast to provide two streams, you need two instances of Mopidy running somewhere, each of which is configured as a separate source for your snapserver instance. You can have a Mopidy-A playing some online radio and Mopidy-B playing some Spotify playlist and then use a controller (e.g. snapdroid) to select which snapclient is playing which of the two streams, or have them both play the same, or mute some/all of them.

Actually, I must be honest. This is how i expect it to work, I don't use it with multiple streams... Maybe someone who does can chime in and be more helpful. I really just wanted to explain why someone might have Mopidy and Snapclient but I got dragged in - forgive me!

**Reply by hillbicks** [reply]
No worries man, I appreciate your answers and I think it comes to down to the same thing as always: Different people have different use cases, and just because my particular use case is not covered by snapcasts functionality, doesn't mean it is not working. It's just not working for me, or so I thought. I just had a discussion with someone on reddit that put me in the right direction. And to be honest, I'm still fairly certain that I'm just not understanding the functionality correctly, although it doesn't seem that I'm the only one. See also #308 

For me at least, muting the clients that I don't want to have playing a stream, just seems to be a workaround. I want to try out the workaround that somebody mentioned on reddit, which is basically creating another source /tmp/snapfifo_off?name=off, which receives no input. Using the API, I can then move clients I want to have playing to the group mopidy, while the others remain in group off. Basically the same thing, but for some reason, this is more straight forward for me. 

I'll try this approach now and I'll report back here in this thread!

Thanks again for your reply

**Reply by kingosticks** [reply]
But one thing that might be worth mentioning: snapserver (by default, unless configured it otherwise) will not send any audio data to "muted" clients. The term "mute" here doesn't just mean the client's volume is turned right down.

**Reply by hillbicks** [reply]
You have no idea HOW stupid I feel, really. This is so embarrassing, it's almost funny. Well, actually it is funny. Mute, just press the damn mute button.

Thanks again dude..........

**Comment by Uatschitchun** [comment]
It's not quite clear what you're trying to achieve. 
I've got a snapserver running on my server streaming Kodi's audio. Kodi is controlled using Yatse. 
Together with that I've got 5 clients (1 being the Server itself, Living Room, Bedroom, Office, Kitchen). They are all in the same one group.
If I like to listen to music in the office and no other room, I mute all but the office. If I like to have music playing all through the house, I unmute all clients. If I'm in the kitchen listening music only there (other clients muted) and like to move over to the living room, I unmute living room and mute kitchen. 
Easy as that. 
If wanting to listen to music in the garden, I use my phone with snapdroid and have a 6th client able to play.

Don't mix "groups" with "scenarios", like predefined groupings of clients playing together! It's not possible with snapcast to group clients into some kind of "listening scenarios", like maybe "cooking" having kitchen and living room as clients or "ambient" having living room and bedroom as clients! A "group" is a set of clients assigned to the same "stream". Ergo: One stream => one group!
(It's possible to design scenarios and I've seen someone doing it. Don't know if you're mentioned article was describing it, too)

I know the HA setup you talked about. Advantage there is, like with my server being setup as snapserver & snapclient simultaneously, so every client is able to play it's own source (own group as it has its own stream) in a kinda "pass-through" way, while also being able to be added to another group (another stream) for acting as a client to that stream. This way the kids are able to play their own music independently from their parents who are listening in living room & kitchen, but are also able to switch over to the parent's stream. 
If there's also a Text-to-Speech stream all clients are grouped into, one is able to also send TTS messages to all clients available, no matter what they are listening to independently. 
If the kids are doing a party, they simply add all clients to one group and have audio throughout the house and all clients playing. 

Does that help?

**Reply by hillbicks** [reply]
Like I just mentioned in the other reply. Apparently this was too obvious for me. And I feel stupid. Like proper stupid. I think I need a drink, a big one. 

Thanks to you as well. This put me in the right direction!

Have a great day

**Comment by tobulli** [comment]
Hi, maybe someone here can help me with the snapcast groups too. I don't understand why it is possible to have multiple groups playing the same stream. Would it not be enough to have one group per stream? Is there a difference in syncing if clients play the same stream but in different groups? I also don't understand why groups change all the time. I would like to use snapcast with Home Assistant and the [mini-media-player](https://github.com/kalkih/mini-media-player) Lovelace card. But it is hard to have a system that always works because of changing group and client names.

Thanks for any help or insight on this :)

**Reply by artmg** [reply]
On your last point, I have shared my lovelace config in #1039 as it's getting off topic here
